### PR TITLE
changed behavior of LastDesignTimeBuildSucceeded following suggestion…

### PR DESCRIPTION
**Pull request by heejaechang**
… on https://github.com/dotnet/roslyn/issues/19626

**Customer scenario**

User opened a solution which contains a project that contains design time build errors. due to that various parts of VS won't work as expected since the project itself is in unknown state. 

for example, "fix all" sometimes doesn't work. errors come and go as project changes its state as design time build get fixed by automatic nuget restore and etc.

this is to reduce confusion on such occasions.

same info is on the bug, but here are new user experience.

"If LastDesignTimeBuildSucceeded=false then

We still get syntax errors from the compiler. (NB: these might not be entirely accurate since we might not have the right language version. I think that’s fine.)
We don’t get semantic errors from the compiler. This means fixes driven from those are broken, but I think that’s fine. Consider “remove unused usings” – most definitely unsafe to invoke.
We still run third party analyzers and code fixes triggered from them. We expect them to respect the flag if needed."

**Bugs this fixes:**

https://github.com/dotnet/roslyn/issues/19626

**Workarounds, if any**

for some cases, user can fix all design time build. some other cases, there is no workaround.

**Risk**

I don't believe this will cause crash since code change is quite straight forward. but this do change existing behavior on this error case. 

**Performance impact**

before, since we dropped projects that have issues, user wouldn't see us doing anything for such broken projects, now we do same work for the broken projects as we have done to good projects. 

but it would be same cost as the same solution without any design time build issues.

**Is this a regression from a previous update?**

NO

**Root cause analysis:**

this is not fixing a code defect. rather we heard some customer feedback and changing our error handling behavior to meet their requests.

**How was the bug found?**

customer report. dogfooding feedback and etc.

### CODE REVIEW

**Comment by heejaechang** [comment]
currently restore doesn't work for me since one of server refuses connections. so I couldn't build yet. so didn't verify fix yet.

**Comment by CyrusNajmabadi** [comment]
Need to understand customre experience here.  If i'm a customer and i just get no semantic errors i'm going to be very confused and unhappy and i'm going to thing VS us just broken.  How do i know what has gone wrong, and what can i do on my end to fix the issue?

**Comment by srivatsn** [comment]
When the designtime build fails there should a more meaningful error in the error list about a missing reference\restore failure etc than the hundreds of missing type errors.

**Comment by heejaechang** [comment]
@srivatsn so who should do that? right now, this is only place which does it

http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/Legacy/AbstractLegacyProject_IIntellisenseBuildTarget.cs,17

and this is only message we currently show
http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/Legacy/AbstractLegacyProject_IIntellisenseBuildTarget.cs,47

but this is only one of case. other case such as missing metadata reference, 
http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Solution/SolutionState.CompilationTracker.cs,638

doesn't output anything to users.

let me know what you guys want on reporting side. the behavior side the PR I linked should do what is described in the description.

**Comment by CyrusNajmabadi** [comment]
How about we just ahve a analyzer/diagnostic that checks if this flag is set.  If it is set we always report a diagnotic to the error list saying that this is the case?

**Comment by sharwell** [comment]
@CyrusNajmabadi It's my understanding that this change did not fix the underlying bug where projects can enter a state where `LastDesignTimeBuildSucceeded` is false but we don't know why. If that's the case, putting a message in the error window regarding this would make projects seem broken even when they are not.

**Comment by srivatsn** [comment]
To answer @heejaechang's question if a design-time build fails, errors from that failure are logged to the errolist with CPS. In the old project system there's one warning that says DT build failed with instructions to get the log (your first link). If there is a missing reference, that's not a DT build failure but the project system outputs a warning that it couldn't resolve the reference.

@CyrusNajmabadi - the first link that HeeJae is pointing at is basically that diagnostic - when DT build fails, it emits that warning.

**Comment by CyrusNajmabadi** [comment]
>  If that's the case, putting a message in the error window regarding this would make projects seem broken even when they are not.

But they are broken.  Because if we've set that bit, then lots of features may have problems.  And hte user will certainly see the results of that problem in that semantic errors and whatnot do not show.

So the choice is:

tell them nothing, and to them the IDE is just broken.

at least tell them *why* we think teh IDE is broken.  It may be that this is bogus on our part, but it's something that can lead to understanding.  i.e. users can then let us know: "this code fix isn't working.  oh, and the error list also says something about the build not working.  but the build did work.  pelase fix this."

As i mentioned in teh last thread, the last time we did this, it took 13 months to track this down.  We need to stop changing feature behavior in a silent fashion.  It is detrimental to customers and to our own ability to track down why things are broken for them.

**Comment by CyrusNajmabadi** [comment]
> If there is a missing reference, that's not a DT build failure but the project system outputs a warning that it couldn't resolve the reference.

Where is this warning?  When we've investigated this issue with customer, there's been nothing in any output window or in the error list.    

Either: 
1. this issue doesn't get reported properly in all cases (and that's bad), or:
2. there's some other issue breaking analysis for customers and we still haven't tracked down the root cause for them.

**Comment by srivatsn** [comment]
Just tried adding a reference to a non-existent dll and I see this:
image

It may be useful to add a diagnostic when LastDesignBuildSucceeded=false and yes there will be duplicate warnings in most cases but it might help us figure out the root cause.

**Comment by heejaechang** [comment]
analyzer seems a good idea. we can add document analyzer that just check the flag and report it if it found a project that has the flag false.

I just need what to write to users in those case.

**Comment by heejaechang** [comment]
@srivatsn so do we want to add analyzer that checks the status of the flag and output diagnostic? if we do, what should I put in the description?

**Comment by davkean** [comment]
Missing or broken references don't break the design-time build by themselves, unless them missing ends up breaking another _custom_ target that expects them to exist.  For example, we have some targets in Roslyn that don't declare their dependencies correctly, which ends up breaking the design-time build. This doesn't happen on a non-customized build.

There's two scenarios that I would like to tackle:

- On opening a CPS-based project, you get hundreds and hundreds of semantic errors (if full analysis is on) until the very first design-time build. We would set this flag in this case, and unset it once we've had a successful build.
- On a broken design-time build, we set this this flag to avoid the user getting hundreds of errors about missing types. 

Currently, when the design-time build is broken in legacy project system, you get the following situation:

```
Warning	IDE0006	Error encountered while loading the project. Some project features, such as full solution analysis for the failed project and projects that depend on it, have been disabled.
Warning		This is an error during design-time
```

Also most often than not (depending on where it's broken) all the source files end up in Miscellaneous Files - where you don't get semantic errors anyway.

In CPS, you get the error that occurred during the design-time build:

```
Error		This is an error during design-time
```
We should consider outputting a similar warning/error that the legacy project system.

All source files also currently end up in Miscellaneous Files (which we're fixing).

@CyrusNajmabadi Based on above, what's your concern?

**Comment by heejaechang** [comment]
@CyrusNajmabadi @srivatsn @sharwell @davkean do we have consent on what to do here?

this should make the behavior to be same as what @srivatsn described in the email. 

For @CyrusNajmabadi concern, I can add something that monitor HasSuccessfullyLoadedAsync (http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Solution/Project.cs,266) and output something in the error list. I can't use diagnostic analyzer since it will only work for open files if FSA is off. but I can add it as incremental analyzer.

I just need what to show users in case "project.HasSuccessfullyLoadedAsync" is false.

...

just to make sure everyone is on same page.

the HasSuccessfuallyLoadedAsync can return false if any of these happened

1. design time build has failed (http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/AbstractProject.cs,312)
2. building skeleton assembly for its cross language p2p reference has failed
(http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Solution/SolutionState.CompilationTracker.cs,630)
3. there is missing metadata references
(http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Solution/SolutionState.CompilationTracker.cs,673)
4. any of its depending p2p reference has HasSuccessfullyLoadedAsync false
(http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Solution/SolutionState.CompilationTracker.cs,686)

basically, if there is anything that can cause cascading semantic errors, that flag/method returns false.

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs [hunk 1]
```diff
@@ -108,13 +108,18 @@ public IEnumerable<DiagnosticData> ConvertToLocalDiagnostics(Document targetDocu
                         }
 
                         // perf optimization. check whether we want to analyze this project or not.
-                        if (!await FullAnalysisEnabledAsync(project, ignoreFullAnalysisOptions, cancellationToken).ConfigureAwait(false))
+                        if (!FullAnalysisEnabled(project, ignoreFullAnalysisOptions))
```
Comment (srivatsn) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs]: Add some tests?

**Comment by davkean** [comment]
I talked to @CyrusNajmabadi offline, and he's in agreement with the design-time build case because we already show something.

With regards to the 2 and 3, what actionable thing can the user do if they fail? I don't know what skeleton assemblies are used for - how personally could I break this if I was a customer and can we provide an actionable fix?

**Comment by heejaechang** [comment]
@davkean 

for 2, it can happen if a project has cross language p2p reference and the referenced project has declaration errors which prevent us from creating metadata only assembly (emit) and there is no previous good assembly to fail over. 
- I have no idea how user would figure this out. we probably need to tell users that "reference project A" contains errors which cause an issue in "project B". so we disabled semantic errors for this project.

for 3, msbuild might show warnings in error list or solution explorer reference node will should (!) if the reference show up there. but it is probably hard for users to connect that dot to understand outcome of that warnings.
- again, I think it would be nice to tell users that "Project A" has missing references "B, C, D" so we disabled semantic errors for the "Project A"

for 4, this is so indirectly affected, even us won't get it right away unless we dig a bit. with all detail messages above + some explicit message such as "Project A" will disable semantic errors since its p2p reference "Project B, C, D" are in broken state.

but these are a lot of information, so not sure even if we show all these information, user can grab all these information.

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs (hunk 1 above)
Comment (heejaechang) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs]: done

File: src/Compilers/CSharp/Portable/CSharpCodeAnalysis.csproj [hunk 2]
```diff
@@ -871,6 +871,7 @@
     <InternalsVisibleToTest Include="Microsoft.CodeAnalysis.Scripting.Destkop.UnitTests" />
     <InternalsVisibleToTest Include="Microsoft.CodeAnalysis.CSharp.Scripting.UnitTests" />
     <InternalsVisibleToTest Include="Microsoft.CodeAnalysis.CSharp.Scripting.Desktop.UnitTests" />
+    <InternalsVisibleToTest Include="Roslyn.Services.Editor.UnitTests" />
```
Comment (heejaechang) [inline src/Compilers/CSharp/Portable/CSharpCodeAnalysis.csproj]: @dotnet/roslyn-compiler can anyone from compiler team can take a look at this change? I added test only IVT so that I can use some internal type in unit tests.

File: src/EditorFeatures/Test/CodeGeneration/CodeGenerationTests.CSharp.cs [hunk 3]
```diff
@@ -342,7 +342,7 @@ public class C
                 var input = "abstract class [|C|] { }";
                 var expected = "abstract class C { public abstract int M(); }";
                 await TestAddMethodAsync(input, expected,
-                    modifiers: new DeclarationModifiers(isAbstract: true),
+                    modifiers: new Editing.DeclarationModifiers(isAbstract: true),
```
Comment (heejaechang) [inline src/EditorFeatures/Test/CodeGeneration/CodeGenerationTests.CSharp.cs]: no test change. added Editing to remove ambiguity.

**Comment by heejaechang** [comment]
pushed change to remove merge conflict from master on test file.

File: src/Compilers/CSharp/Portable/CSharpCodeAnalysis.csproj (hunk 2 above)
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/CSharpCodeAnalysis.csproj]: IVT to test assemblies is okay.

**Comment by heejaechang** [comment]
I opened an issue to separate notification part out of this bug. I can work on the notification part once people decide what/how to show message to users. I already listed all cases I know which project can fail to load successfully above.

**Comment by heejaechang** [comment]
@dotnet/roslyn-infrastructure failure doesn't look like related to my change.

https://ci.dot.net/job/Private/job/dotnet_roslyn-internal/job/master/job/windows_eta_open_prtest/6752/

..
16:20:09 System.IO.FileLoadException: Could not load file or assembly 'Microsoft.VisualStudio.Validation, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
16:20:09 File name: 'Microsoft.VisualStudio.Validation, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
16:20:09    at Roslyn.Hosting.Test.StartTargetAction.<>c.<StartApp>b__0_0()
16:20:09    at Roslyn.Hosting.Test.EditorTestAppAction.<>c__DisplayClass25_0.<ExecuteOnUIThread>b__0() in D:\j\workspace\windows_eta_o---d880a2eb\Closed\Hosting\Test\Actions\EditorTestApp\EditorTestAppAction.cs:line 165
16:20:09 --- End of stack trace from previous location where exception was thrown ---
16:20:09    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
16:20:09    at Roslyn.Hosting.Test.EditorTestAppAction.ExecuteOnUIThread(Action a) in D:\j\workspace\windows_eta_o---d880a2eb\Closed\Hosting\Test\Actions\EditorTestApp\EditorTestAppAction.cs:line 178
16:20:09    at Roslyn.Hosting.Test.StartTargetAction.StartApp() in D:\j\workspace\windows_eta_o---d880a2eb\Closed\Hosting\Test\Actions\EditorTestApp\StartTargetAction.cs:line 19
16:20:09    at Roslyn.Hosting.Test.Runner.<RunInBackground>b__7_0() in D:\j\workspace\windows_eta_o---d880a2eb\Closed\Hosting\Test\Execution\EditorTestApp\Runner.cs:line 73
..

**Comment by jasonmalinowski** [comment]
@rchande is fixing that right now.

**Comment by heejaechang** [comment]
retest windows_eta_open_prtest please

**Comment by heejaechang** [comment]
retest windows_eta_open_prtest  please

**Comment by heejaechang** [comment]
retest windows_release_vs-integration_prtest please

**Comment by heejaechang** [comment]
tagging @MattGertz for 15.3 approval

**Comment by MattGertz** [comment]
MoreInfo -- I need some more context on this one:
@srivatsn what was the initial motivation for #19626 ? I see it was apparently Customer feedback->Dogfooding, but what sort of volume are we talking about here?  Behavioral changes like this make me nervous -- have we run this solution by PM or impacted customers to get a sense as to whether or not it really solves the problem?

@heejaechang Not quite understanding the comment on filing separate issues for the notifications.  Does that mean that, for some period of time, we will be in the situation that @CyrusNajmabadi is correctly concerned about v.v. changing behavior silently?  At the very least, I'd want customer notification to accompany this change.

**Comment by Pilchie** [comment]
Jumping in - since I happened to be looking at this PR anyway:

For the second bullet, we're *already* in that state, except that as well as not getting errors, you also don't get lightbulbs, fix-all, etc.  This is a definite improvement in the experience, though we can (and should) go further to let the user know what's going on.

**Comment by heejaechang** [comment]
@Pilchie @MattGertz Kevin, already said it well. so I might not need to add more. basically notification wise, nothing changed. we already don't tell users about the situation in some cases. and I agree it would be nice if we can add more notification for users.

**Comment by MattGertz** [comment]
OK, thanks.  Still looking for more details on current customer pain.

**Comment by srivatsn** [comment]
The motivations for this are:
- xproj parity: Creating a project today shows a bunch of squiggles in the IDE before background restore completes and they go away. [Fowler says](https://github.com/dotnet/project-system/issues/928#issuecomment-266636521) this was one of the big things they had to do in xproj. We've seen some vsfeedback about this as well (I can try to dig them up)
- When a design-time build fails, these errors muddy the error list and the real errors get buried. So all the recent vsfeedback we've seen around random errors in the error list could be related to this - where there was a real design-time build failure (but not a build failure) and the user has no clue how to debug through it. For this one, it's hard to get data because, we have a lot of vsfeedback about errors in error list but not in build and it's hard to know which of those are caused by this. We've fixed a bunch of other known causes and if we fix this, then we can get to real reasons of failures instead of just missing system.object.

**Comment by MattGertz** [comment]
OK.  I would have problems with this in Escrow, but I am approving given we've got a bit more runway to validate.

**Comment by heejaechang** [comment]
ETA test failure is known issue.
